Dynamo comptime, with public ComptimeContext API

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90983

This PR adds `@comptime`, a decorator that causes a given function to be executed at compile time when Dynamo is symbolically evaluating their program. To query the Dynamo state, we offer a public ComptimeContext API which provides a limited set of APIs for querying Dynamo's internal state. We intend for users to use this API and plan to keep it stable. Here are some things you can do with it:

* You want to breakpoint Dynamo compilation when it starts processing a particular line of user code: give comptime a function that calls breakpoint
* You want to manually induce a graph break for testing purposes; give comptime a function that calls unimplemented
* You want to perform a debug print, but you don't want to induce a graph break; give comptime a function that prints.
* You can print what the symbolic locals at a given point in time are.
* You can print out the partial graph the Dynamo had traced at this point.
* (My original motivating use case.) You want to add some facts to the shape env, so that a guard evaluation on an unbacked SymInt doesn't error with data-dependent. Even if you don't know what the final user API for this should be, with comptime you can hack out something quick and dirty. (This is not in this PR, as it depends on some other in flight PRs.)

Check out the tests to see examples of comptime in action.

In short, comptime is a very powerful debugging tool that lets you drop into Dynamo from user code, without having to manually jerry-rig pdb inside Dynamo to trigger after N calls.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

ezyang committed:
[RFC] Dynamo comptime

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

voznesenskym commented:
Brief thoughts, discussed offline somewhat. 

Philosophically, I think this is a great tool. This is v. similar to what I mentioned in the NFC chat (cc @Chillee @zou3519) w/r/t to giving users a VariableTracker, hooked into the builder, to populate as a compile time alternative to help interpret their types instead of just/only using @allow_in_graph. 

Implementation wise - exposing tx, output_graph, and other core internals is going to be an extremely difficult contract to maintain. It essentially broadens the entire surface area of what we have to consider as "user facing" - to the entirety of dynamo. If we do go ahead with this, I think we need to either:

A) Expose a clear subset API like `OutputGraphPublic`, `InstructionTranslatorPublic`, where `InstructionTranslator` is now `InstructionTranslator(InstructionTranslatorPublic)` and type-erase that down to just `InstructionTranslatorPublic` when we pass it back:

```
class InstructionTranslatorPublic:

@property
def output() -> OutputGraphPublic:

class InstructionTranslator(InstructionTranslatorPublic):

@property
def _output() -> OutputGraph
```

This will be annoying to maintain, and add a lot of class structures, but will also signal to users what is and is not a real API, as well as force us to think about what is, and is not, okay to expose or contractualize

B) Do what you did in the PR, but give it extremely clear signaling around unsafety, like `comptime_not_back_compat_ever` or `comptime_unsafe` or something like that, and always make sure the contract is `*args/**kwargs` so even new args or removal of old ones doesnt break contract.

mlazos commented:
I'm really happy with this idea and would definitely use this, mainly as a debugging tool. I think your suggestion about having this be an underlying API which we narrowly expose to users with small public facing APIs is a good idea to address the concern of exposing all of dynamo as a public contract.

jansel reviewed (commented):
I'm ok with this idea as a debugging tool, though I wouldn't want to maintain it as a BC surface.  So perhaps add some sort of giant warning on it.

ezyang committed:
Update on "[RFC] Dynamo comptime"


The purpose of this PR is to spark a discussion about whether or not we want this feature or not. I do not care if we throw out this PR entirely, it took very little time to prepare. I'm curious what people think about it.

This PR adds a special `comptime` function to Dynamo. This function takes a user defined function, and runs that function at Dynamo compile time (with access to the current InstructionTranslator). It does nothing otherwise. Comptime is very, very flexible (perhaps too flexible). Here are some things you can do with it:

* You want to breakpoint Dynamo compilation when it starts processing a particular line of user code: give comptime a function that calls breakpoint
* You want to manually induce a graph break for testing purposes; give comptime a function that calls unimplemented
* You want to perform a debug print, but you don't want to induce a graph break; give comptime a function that prints. You can access InstructionTranslator to poke out information about the symbolic locals at this point in time. You could also print out the partial graph the Dynamo had traced at this point.
* You want to do template metaprogramming in Python? Sure. Just munge up the instruction translator state as necessary.
* (My original motivating use case.) You want to add some facts to the shape env, so that a guard evaluation on an unbacked SymInt doesn't error with data-dependent. Even if you don't know what the final user API for this should be, with comptime you can hack out something quick and dirty.

In short, comptime is a very powerful tool that lets you build extensions on top of Dynamo, without having to actually modify the Dynamo codebase.

Why should we NOT do this? The main reason I can think of is that it exposes all of Dynamo's internals (e.g., InstructionTranslator, VariableTracker, etc). If people start relying on this API, it impedes our ability to do refactors in the future. We could mitigate this by providing a narrow public API which you have to go through to get the goodies. E.g., I would add APIs for shape env manipulation here.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
This PR has been updated with a proposal API for the public bindings that allow you to access Dynamo internal state. I took great pains to reduce the number of commitments we make to future users, while making the API flexible enough to allow you to do most things you might want to do when debugging. Check it out!

ezyang committed:
Update on "Dynamo comptime, with public ComptimeContext API"


This PR adds `comptime`, a decorator that causes a given function to be executed at compile time when Dynamo is symbolically evaluating their program. To query the Dynamo state, we offer a public ComptimeContext API which provides a limited set of APIs for querying Dynamo's internal state. We intend for users to use this API and plan to keep it stable. Here are some things you can do with it:

* You want to breakpoint Dynamo compilation when it starts processing a particular line of user code: give comptime a function that calls breakpoint
* You want to manually induce a graph break for testing purposes; give comptime a function that calls unimplemented
* You want to perform a debug print, but you don't want to induce a graph break; give comptime a function that prints.
* You can print what the symbolic locals at a given point in time are.
* You can print out the partial graph the Dynamo had traced at this point.
* (My original motivating use case.) You want to add some facts to the shape env, so that a guard evaluation on an unbacked SymInt doesn't error with data-dependent. Even if you don't know what the final user API for this should be, with comptime you can hack out something quick and dirty. (This is not in this PR, as it depends on some other in flight PRs.)

Check out the tests to see examples of comptime in action.

In short, comptime is a very powerful debugging tool that lets you drop into Dynamo from user code, without having to manually jerry-rig pdb inside Dynamo to trigger after N calls.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Dynamo comptime, with public ComptimeContext API"


This PR adds `comptime`, a decorator that causes a given function to be executed at compile time when Dynamo is symbolically evaluating their program. To query the Dynamo state, we offer a public ComptimeContext API which provides a limited set of APIs for querying Dynamo's internal state. We intend for users to use this API and plan to keep it stable. Here are some things you can do with it:

* You want to breakpoint Dynamo compilation when it starts processing a particular line of user code: give comptime a function that calls breakpoint
* You want to manually induce a graph break for testing purposes; give comptime a function that calls unimplemented
* You want to perform a debug print, but you don't want to induce a graph break; give comptime a function that prints.
* You can print what the symbolic locals at a given point in time are.
* You can print out the partial graph the Dynamo had traced at this point.
* (My original motivating use case.) You want to add some facts to the shape env, so that a guard evaluation on an unbacked SymInt doesn't error with data-dependent. Even if you don't know what the final user API for this should be, with comptime you can hack out something quick and dirty. (This is not in this PR, as it depends on some other in flight PRs.)

Check out the tests to see examples of comptime in action.

In short, comptime is a very powerful debugging tool that lets you drop into Dynamo from user code, without having to manually jerry-rig pdb inside Dynamo to trigger after N calls.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Dynamo comptime, with public ComptimeContext API"


This PR adds `comptime`, a decorator that causes a given function to be executed at compile time when Dynamo is symbolically evaluating their program. To query the Dynamo state, we offer a public ComptimeContext API which provides a limited set of APIs for querying Dynamo's internal state. We intend for users to use this API and plan to keep it stable. Here are some things you can do with it:

* You want to breakpoint Dynamo compilation when it starts processing a particular line of user code: give comptime a function that calls breakpoint
* You want to manually induce a graph break for testing purposes; give comptime a function that calls unimplemented
* You want to perform a debug print, but you don't want to induce a graph break; give comptime a function that prints.
* You can print what the symbolic locals at a given point in time are.
* You can print out the partial graph the Dynamo had traced at this point.
* (My original motivating use case.) You want to add some facts to the shape env, so that a guard evaluation on an unbacked SymInt doesn't error with data-dependent. Even if you don't know what the final user API for this should be, with comptime you can hack out something quick and dirty. (This is not in this PR, as it depends on some other in flight PRs.)

Check out the tests to see examples of comptime in action.

In short, comptime is a very powerful debugging tool that lets you drop into Dynamo from user code, without having to manually jerry-rig pdb inside Dynamo to trigger after N calls.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jansel commented on test/dynamo/test_comptime.py:
```diff
@@ -0,0 +1,216 @@
+
+            @comptime
+            def _(ctx):
+                ctx.print_disas(file=FILE)
```
Perhaps we should expose some of these through a more compact API like `dynamo.debug.print_disas()` (which would just be this same function) -- I'd expect a lot of common cases will just be calling a single one of these functions.

ezyang replied:
The one subtlety is that we would have to make all of these functions special forms in Dynamo, in the same way `comptime` has special handling. We can still do that though. Another possibility is to just define the function as

```
def print_disas(ctx):
    ctx.print_disas()
```

and then you can simply say `comptime(torch._dynamo.debug.print_disas)` to do the single line, without having to introduce other special forms (but you'd still have to import modules for two identifiers.) These are both easy to implement so it's mostly a matter of what people prefer.

ezyang replied:
Actually, I guess if you don't mind introducing a stack frame (which does matter in a few cases), you can just use functional abstraction. I'll go add these.

ezyang commented on torch/_dynamo/comptime.py:
```diff
@@ -0,0 +1,219 @@
+            file=file,
+        )
+
+    def _i_will_not_complain_if_bc_breaks_InstructionTranslator(self):
```
Ah but you see, the glorious thing is that double underscore prefixed identifiers in Python get name mangled, so you'd have to write `ctx._ComptimeContext__tx` anyway! Take that! (Well, I guess the mangled name is still shorter though haha)

ezyang committed:
Update on "Dynamo comptime, with public ComptimeContext API"


This PR adds `comptime`, a decorator that causes a given function to be executed at compile time when Dynamo is symbolically evaluating their program. To query the Dynamo state, we offer a public ComptimeContext API which provides a limited set of APIs for querying Dynamo's internal state. We intend for users to use this API and plan to keep it stable. Here are some things you can do with it:

* You want to breakpoint Dynamo compilation when it starts processing a particular line of user code: give comptime a function that calls breakpoint
* You want to manually induce a graph break for testing purposes; give comptime a function that calls unimplemented
* You want to perform a debug print, but you don't want to induce a graph break; give comptime a function that prints.
* You can print what the symbolic locals at a given point in time are.
* You can print out the partial graph the Dynamo had traced at this point.
* (My original motivating use case.) You want to add some facts to the shape env, so that a guard evaluation on an unbacked SymInt doesn't error with data-dependent. Even if you don't know what the final user API for this should be, with comptime you can hack out something quick and dirty. (This is not in this PR, as it depends on some other in flight PRs.)

Check out the tests to see examples of comptime in action.

In short, comptime is a very powerful debugging tool that lets you drop into Dynamo from user code, without having to manually jerry-rig pdb inside Dynamo to trigger after N calls.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
Updated:
- API now supports stacklevel so you can parent outer InstructionTranslator
- Convenience API comptime.print_X() now exists, for compact debugging.

ezyang commented:
@pytorchbot merge

ezyang commented on torch/_dynamo/variables/misc.py:
```diff
@@ -87,6 +88,61 @@ class UnknownVariable(VariableTracker):
+                # a free variable that we actually DO have the runtime
+                # value for
+                # tuple(make_cell(ComptimeVar(i)) for i in fn.closure.items)
+                tuple(),
```
So, automatically promoting free variables into ComptimeVar is hella confusing, so we shouldn't do it, but I also thought about automatically calling `as_python_constant` on free variables. The problem is I can't guarantee that object identity holds (e.g., you access a "constant" list, that list will actually get reconstructed) and then mutations don't work correctly. Also nonlocal handling with cells is a pain.
